### PR TITLE
Balder/add mmappool

### PR DESCRIPTION
**Pull request by baldersheim**
@vekterli PR

### CODE REVIEW

**Review by vekterli — APPROVED** [review]
:+1: just some questions

File: vespamalloc/src/vespamalloc/malloc/threadpool.hpp [hunk 1]
```diff
@@ -71,13 +71,19 @@ mallocHelper(size_t exactSize,
                 PARANOID_CHECK2( *(int *)2 = 2; );
             }
         } else {
-            af._allocFrom = _allocPool->exactAlloc(exactSize, sc, af._allocFrom);
-            _stat[sc].incExactAlloc();
-            if (af._allocFrom) {
-                af._allocFrom->sub(mem);
-                PARANOID_CHECK2( if (!mem.ptr()) { *(int *)3 = 3; } );
+            if (exactSize > _mmapLimit) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespamalloc/src/vespamalloc/malloc/threadpool.hpp:74]: Added capping between 1M and 1G.
The allocation is always aligned and rounded to huge page size.

File: vespamalloc/src/vespamalloc/malloc/mmappool.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        MMapInfo(size_t id, size_t sz) : _id(id), _sz(sz) { }
+        size_t _id;
+        size_t _sz;
+    };
+    const size_t        _page_size;
+    const int           _huge_flags;
+    std::atomic<size_t> _count;
+    std::atomic<bool>   _has_hugepage_failure_just_happened;
+    mutable std::mutex  _mutex;
+    std::map<const void *, MMapInfo> _mappings;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespamalloc/src/vespamalloc/malloc/mmappool.h:30]: Yes, done

File: vespamalloc/src/vespamalloc/malloc/threadpool.hpp [hunk 3]
```diff
@@ -71,13 +71,19 @@ mallocHelper(size_t exactSize,
...
+                mem = MemBlockPtrT(_mmapPool->mmap(MemBlockPtrT::classSize(sc)), MemBlockPtrT::classSize(sc));
+                mem.setExact(exactSize);
+                mem.free();
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespamalloc/src/vespamalloc/malloc/threadpool.hpp:77]: Comment added, the sanity check is conducted at the call site.

**Pull request merged**